h2o-core/src/main/java/hex/PartialDependence.java

```diff
@@ -94,151 +116,208 @@ private void checkSanityAndFillParams() {
-                      meanResponse[which] = calMeansSTD.getWeightedMean();;
-                      stddevResponse[which] = calMeansSTD.getWeightedSigma();
-                      stdErrorOfTheMeanResponse[which] = stddevResponse[which] / Math.sqrt(preds.numRows());
-                    } else {
-                      meanResponse[which] = preds.vec(2).mean();
-                      stddevResponse[which] = preds.vec(2).sigma();
-                      stdErrorOfTheMeanResponse[which] = stddevResponse[which] / Math.sqrt(preds.numRows());
-                    }
-                  } else if (_model_id.get()._output.nclasses() == 1) {
-                    if (_weight_column_index >= 0) {
-                      CalculateWeightMeanSTD calMeansSTD = getWeightedStat(fr, preds, 0);
-                      meanResponse[which] = calMeansSTD.getWeightedMean();;
-                      stddevResponse[which] = calMeansSTD.getWeightedSigma();
-                      stdErrorOfTheMeanResponse[which] = stddevResponse[which] / Math.sqrt(preds.numRows());
-                    } else {
-                      meanResponse[which] = preds.vec(0).mean();
-                      stddevResponse[which] = preds.vec(0).sigma();
-                      stdErrorOfTheMeanResponse[which] = stddevResponse[which] / Math.sqrt(preds.numRows());
-                    }
-                  } else throw H2O.unimpl();
+                  fr = _frame_id.get();
                 }
-              } finally {
-                if (preds != null) preds.remove();
-              }
-              cons.remove();
-              if (_row_index > 0) {
-                fr.remove();
+                Frame test = new Frame(fr.names(), fr.vecs());
+                Vec orig = test.remove(col);
+                Vec cons = orig.makeCon(value);
+                if (cat) cons.setDomain(fr.vec(col).domain());
+                test.add(col, cons);
+                
+                Vec cons2 = null;
+                if (workOn2D) {
+                  Vec orig2 = test.remove(col2);
+                  cons2 = orig2.makeCon(value2);
+                  if (cat2) cons2.setDomain(fr.vec(col2).domain());
+                  test.add(col2, cons2);
+                }
+                Frame preds = null;
+                try {
+                  preds = _model_id.get().score(test, Key.make().toString(), _job, false);
+                  if (preds == null || preds.numRows() == 0) {  // this can happen if algo will not predict on rows with NAs
+                    meanResponse[which] = Double.NaN;
+                    stddevResponse[which] = Double.NaN;
+                    ;
+                    stdErrorOfTheMeanResponse[which] = Double.NaN;
+                    ;
+                  } else {
+                    if (_model_id.get()._output.nclasses() == 2) {
+                      if (_weight_column_index >= 0) { // calculated weighted statistics
```

Comment:
i think we are here about four leves of nested blocks too deep

Reply:
So picky!  Anyway, I tried to fix it.  Please check again and let me know.  Thanks, Wendy